editors/code/src/command/workspace-folder-formatting.ts

```diff
@@ -0,0 +1,228 @@
+ * - The LSP server part of Air needs to perfectly track changes for any open
+ *   files. Closing the tabs first switches the server's "source of truth" from
+ *   the client version of the file to the on-disk version of the file, which
+ *   should ensure the LSP server won't have any chance of getting out of sync.
+ *   This may be overkill, since the LSP server seems to handle git branch
+ *   changes well, but it's still not a bad idea.
+ *
+ * # Forward propagation of settings
+ *
+ * If no `air.toml` is tied to a document, then typically in the LSP document
+ * settings are forward propagated from the client to the server, and those are
+ * used during formatting when that document is open. If an `air.toml` does
+ * exist, those settings override document settings.
+ *
+ * Because we go through the Air CLI here, forward propagated document settings
+ * are ignored, so even if you don't have an `air.toml`, you will get Air's
+ * default indentation settings.
+ *
+ * We can't simply require an `air.toml` at the workspace folder root to avoid
+ * confusion here, because you may instead have an `air.toml` at
+ * `{root}/R/air.toml` that would handle any R files in your project. Instead,
+ * we hope this isn't common enough to come up much in practice.
+ */
+export function workspaceFolderFormatting(ctx: Ctx): Cmd {
+	return async () => {
+		const binaryPath = ctx.lsp.getBinaryPath();
+
+		const workspaceFolder = await selectWorkspaceFolder();
+		if (!workspaceFolder) {
+			return;
+		}
+
+		const allTabsClosed = await closeAllTabs(workspaceFolder);
+		if (!allTabsClosed) {
+			return;
+		}
+
+		const workspaceFolderPath = workspaceFolder.uri.fsPath;
+
+		let stderr = "";
+		let anyErrors = false;
+
+		// i.e., `air format {workspaceFolderPath} --no-color`
+		const args = ["format", workspaceFolderPath, "--no-color"];
+
+		// This should not matter since the path is explicitly supplied, but better to be safe
+		const options = {
+			cwd: workspaceFolderPath,
+		};
+
+		// A promise that resolves when the spawned process closes or errors
+		const finishedFormatting = new Promise<void>((resolve) => {
+			// Use spawn instead of exec to avoid maxBufferExceeded error
+			const p = cp.spawn(binaryPath, args, options);
+
+			p.stderr.setEncoding("utf8");
+			p.stderr.on("data", (data) => (stderr += data));
+
+			p.on("error", () => {
+				anyErrors = true;
+				return resolve();
+			});
+
+			p.on("close", (code) => {
+				if (code !== 0) {
+					anyErrors = true;
+				}
+				return resolve();
+			});
+		});
+
+		await finishedFormatting;
+
+		if (anyErrors) {
+			output.log(
+				`Errors occurred while formatting the ${workspaceFolder.name} workspace folder.\n${stderr}`,
+			);
+
+			const answer = await vscode.window.showInformationMessage(
+				`Errors occurred while formatting the ${workspaceFolder.name} workspace folder. View the logs?`,
+				{ modal: true },
+				"Yes",
+				"No",
+			);
+
+			if (answer === "Yes") {
+				output.show();
+			}
```

Comment:
I didn't show it, but if any files fail to format we write the emitted stderr into the logs and offer to take the user there

Reply:
Nice!